File: services/worker-go/worker/environment.go
package main

// CompiledEnvironment is the Go worker configuration baked into the binary.
//
// Edit the values below before building the Go worker. Unlike the Python
// worker, the Go worker intentionally does not read configuration from process
// environment variables or .env files at runtime.
//
// Never commit production credentials to a public repository. Build production
// binaries from a private copy of this file and publish the resulting image to
// a private registry.
type CompiledEnvironment struct {
	FirebaseDatabaseURL string
	ServiceAccountJSON  string
	WorkspaceID         string
	PoolID              string
	WorkerID            string
	WorkerMachineID     string
	WorkerToken         string
	WorkerLabel         string
	WorkerLocation      string
	WorkerShards        []string
	QueueShardCount     int
	MaxConcurrency      int
	LeaseSeconds        int
	PollSeconds         int
	CloneDir            string
	DataDir             string
	EncryptionKey       string
	NgrokEnabled        bool
	NgrokAuthtoken      string
	NgrokBin            string
	NgrokRegion         string
}

var Environment = CompiledEnvironment{
	FirebaseDatabaseURL: "",
	ServiceAccountJSON:  "",
	WorkspaceID:         "default",
	PoolID:              "default",
	WorkerID:            "",
	WorkerMachineID:     "",
	WorkerToken:         "",
	WorkerLabel:         "",
	WorkerLocation:      "",
	WorkerShards:        nil,
	QueueShardCount:     16,
	MaxConcurrency:      2,
	LeaseSeconds:        90,
	PollSeconds:         5,
	CloneDir:            "/app/clones",
	DataDir:             "/app/data",
	EncryptionKey:       "",
	NgrokEnabled:        false,
	NgrokAuthtoken:      "",
	NgrokBin:            "ngrok",
	NgrokRegion:         "",
}

File: services/worker-rust/src/environment.rs
#[allow(dead_code)]
#[derive(Debug, Clone)]
pub struct CompiledEnvironment {
    pub firebase_database_url: &'static str,
    pub service_account_json: &'static str,
    pub workspace_id: &'static str,
    pub pool_id: &'static str,
    pub worker_id: &'static str,
    pub worker_machine_id: &'static str,
    pub worker_token: &'static str,
    pub worker_label: &'static str,
    pub worker_location: &'static str,
    pub worker_shards: Vec<String>,
    pub queue_shard_count: u64,
    pub max_concurrency: u64,
    pub lease_seconds: u64,
    pub poll_seconds: u64,
    pub clone_dir: &'static str,
    pub data_dir: &'static str,
    pub encryption_key: &'static str,
    pub ngrok_enabled: bool,
    pub ngrok_authtoken: &'static str,
    pub ngrok_bin: &'static str,
    pub ngrok_region: &'static str,
    pub configured_at_build: bool,
}

include!(concat!(env!("OUT_DIR"), "/environment_generated.rs"));

pub fn validate(environment: &CompiledEnvironment) -> Result<(), String> {
    if environment.firebase_database_url.trim().is_empty() {
        return Err("FirebaseDatabaseURL is empty in the compiled Rust environment".into());
    }
    if environment.service_account_json.trim().is_empty() {
        return Err("ServiceAccountJSON is empty in the compiled Rust environment".into());
    }
    let account: serde_json::Value = serde_json::from_str(environment.service_account_json)
        .map_err(|error| format!("ServiceAccountJSON is invalid: {error}"))?;
    if account.get("client_email").and_then(|value| value.as_str()).unwrap_or("").is_empty()
        || account.get("private_key").and_then(|value| value.as_str()).unwrap_or("").is_empty()
    {
        return Err("ServiceAccountJSON must include client_email and private_key".into());
    }
    if environment.encryption_key.trim().is_empty() {
        return Err("EncryptionKey is empty in the compiled Rust environment".into());
    }
    Ok(())
}
